map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java

```diff
@@ -0,0 +1,415 @@
+          new Alert(Alert.AlertType.ERROR, "Failed to get the Preplanned Map Areas from the Offline Map Task.").show();
+        }
+      });
+
+      // add a cell factory to the list view that shows the preplanned area's title
+      preplannedAreasListView.setCellFactory(c -> new PreplannedMapAreaListCell());
+
+      // create a listener that changed the viewpoint to the extent of the preplanned map area when an item is selected in the list view
+      listViewSelectionChangeViewpointListener = (obs, oldValue, newValue) -> {
+        if (newValue != null) {
+          // get the extent of the area of interest, and add a buffer
+          Envelope areaOfInterest = GeometryEngine.buffer(newValue.getAreaOfInterest(), 100).getExtent();
+
+          // set the point of view to the area of interest
+          mapView.setViewpointAsync(new Viewpoint(areaOfInterest), 1.50F);
+        }
+      };
+
+      // create a listener that checks if the area has been previously downloaded, and changes the button text accordingly
+      ChangeListener<PreplannedMapArea> listViewSelectionButtonLabelListener = (obs, oldValue, newValue) -> {
+        if (newValue != null) {
+          // get the path for the preplanned area
+          String path = tempDirectory + "/" + newValue.getPortalItem().getTitle();
+
+          // check if the path exists and set the button text accordingly
+          if (Files.exists(Paths.get(path))) {
+            downloadAreaButton.setText("View Downloaded Area");
+          } else {
+            downloadAreaButton.setText("Download Preplanned Area");
+          }
+        }
+      };
+
+      // add the listeners to the list view
+      preplannedAreasListView.getSelectionModel().selectedItemProperty().addListener(listViewSelectionChangeViewpointListener);
+      preplannedAreasListView.getSelectionModel().selectedItemProperty().addListener(listViewSelectionButtonLabelListener);
+
+    } catch (IOException e) {
+      new Alert(Alert.AlertType.ERROR, "Failed to create a temporary path for saving the Preplanned Map Areas.").show();
+    }
+  }
+
+  /**
+   * Downloads the selected preplanned map area.
+   */
+  @FXML
+  private void handleDownloadArea() {
+    if (preplannedAreasListView.getSelectionModel().getSelectedItem() != null) {
+
+      // remove the listener from the list view to stop the viewpoint changing on selection
+      preplannedAreasListView.getSelectionModel().selectedItemProperty().removeListener(listViewSelectionChangeViewpointListener);
+
+      // enable the 'view web map' button
+      showWebMapButton.setDisable(false);
+
+      // get the requested preplanned map area
+      PreplannedMapArea selectedMapArea = preplannedAreasListView.getSelectionModel().getSelectedItem();
+
+      // create a folder path where the map package will be downloaded to
+      String path = tempDirectory + "/" + selectedMapArea.getPortalItem().getTitle();
+
+      // if the area is already downloaded, open it
```

Comment:
I find it makes sense to check whether the area has already been downloaded at the very beginning though, to avoid downloading it again. We could flip it around and check whether there _arent_ any files:
```
if (!Files.exists(Paths.get(path))) {
 // download the map
} else {
 // open the map
}
```
Although I'd have to think about the implications of this.

Reply:
Turned this statement around as described below. Definitely reads nicer when the block for downloading appears first. Some minor refactoring as well as a result. Let me know what you think.